homeassistant/components/smartthings/climate.py

```diff
@@ -0,0 +1,226 @@
+        super().__init__(device)
+        self._base_supported_features = self._determine_base_features()
+        self._supported_features = 0
+
+    def _determine_base_features(self):
+        from pysmartthings import Capability
+
+        flags = SUPPORT_OPERATION_MODE
+        if self._device.get_capability(
+                Capability.thermostat_fan_mode, Capability.thermostat):
+            flags |= SUPPORT_FAN_MODE
+        return flags
+
+    async def async_set_fan_mode(self, fan_mode):
+        """Set new target fan mode."""
+        await self._device.set_thermostat_fan_mode(fan_mode, set_status=True)
+
+        # State is set optimistically in the command above, therefore update
+        # the entity state ahead of receiving the confirming push updates
+        self.async_schedule_update_ha_state(True)
+
+    async def async_set_operation_mode(self, operation_mode):
+        """Set new target operation mode."""
+        mode = STATE_TO_MODE[operation_mode]
+        await self._device.set_thermostat_mode(mode, set_status=True)
+
+        # State is set optimistically in the command above, therefore update
+        # the entity state ahead of receiving the confirming push updates
+        self.async_schedule_update_ha_state(True)
+
+    async def async_set_temperature(self, **kwargs):
+        """Set new operation mode and target temperatures."""
+        # Operation state
+        operation_state = kwargs.get(ATTR_OPERATION_MODE)
+        if operation_state:
+            mode = STATE_TO_MODE[operation_state]
+            await self._device.set_thermostat_mode(mode, set_status=True)
+
+        # Heat/cool setpoint
+        heating_setpoint = None
+        cooling_setpoint = None
+        if self.current_operation == STATE_HEAT:
+            heating_setpoint = kwargs.get(ATTR_TEMPERATURE)
+        elif self.current_operation == STATE_COOL:
+            cooling_setpoint = kwargs.get(ATTR_TEMPERATURE)
+        else:
+            heating_setpoint = kwargs.get(ATTR_TARGET_TEMP_LOW)
+            cooling_setpoint = kwargs.get(ATTR_TARGET_TEMP_HIGH)
+        tasks = []
+        if heating_setpoint is not None:
+            tasks.append(self._device.set_heating_setpoint(
+                round(heating_setpoint, 3), set_status=True))
+        if cooling_setpoint is not None:
+            tasks.append(self._device.set_cooling_setpoint(
+                round(cooling_setpoint, 3), set_status=True))
+        await asyncio.gather(*tasks)
+
+        # State is set optimistically in the commands above, therefore update
+        # the entity state ahead of receiving the confirming push updates
+        self.async_schedule_update_ha_state(True)
+
+    async def async_update(self):
+        """Update the device."""
+        # Flags need to be set dynamically based on operation mode.
+        flags = self._base_supported_features
+        if self.current_operation == STATE_AUTO:
+            flags |= SUPPORT_TARGET_TEMPERATURE_LOW \
```

Comment:
Supported features should not change during entity lifetime.

Reply:
Unfortunately, this is expected for the climate entity--it's a mess and won't function correctly otherwise.  I can add a more verbose comment calling out this anti-pattern, but there's really no way around it right now until #home-assistant/architecture/issues/22 is resolved.  While precedent is not a reason to continue bad practices, this is how the other thermostat climate platforms operate as well.

Reply:
We shouldn't make decisions in the backend to cater for the frontend not being smart.

I looked through all the climate platforms now. None of the platforms that use high and low temps changed the supported features regarding these during entity lifetime. Looks at ecobee, elkm1, mysensors, nest, wink eg.

Reply:
Your right!  I don’t know what I was looking at...  I’ll have to spend some more time figuring out why it doesn’t behave the same as Nest with static flags.